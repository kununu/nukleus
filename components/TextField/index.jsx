// Read more about controlled components
// https://facebook.github.io/react/docs/forms.html#controlled-components
import React, {Component, PropTypes} from 'react';

import styles from './index.scss';

import {
  controlLabel,
  controlNote,
  errorStyles,
  hidden,
  formControl,
  formGroup,
  srOnly,
  controlLabelRequired
} from '../index.scss';

export default class TextField extends Component {
  static propTypes = {
    autoComplete: PropTypes.string,
    autoFocus: PropTypes.bool,
    disable: PropTypes.bool,
    error: PropTypes.string,
    id: PropTypes.string.isRequired,
    inputStyle: PropTypes.string,
    isRequired: PropTypes.bool,
    label: PropTypes.string.isRequired,
    labelHidden: PropTypes.bool,
    maxLength: PropTypes.number,
    multiLine: PropTypes.bool,
    name: PropTypes.string.isRequired,
    onChange: PropTypes.func,
    pattern: PropTypes.string,
    placeholder: PropTypes.string,
    query: PropTypes.object,
    requiredLabel: PropTypes.string,
    rows: PropTypes.number,
    title: PropTypes.string,
    type: PropTypes.oneOf([
      'email',
      'password',
      'text',
      'url'
    ]),
    value: PropTypes.string
  };

  static defaultProps = {
    autoComplete: 'off',
    autoFocus: false,
    disable: false,
    error: null,
    inputStyle: 'inline',
    isRequired: false,
    labelHidden: false,
    maxLength: 500,
    multiLine: false,
    onChange: null,
    pattern: '',
    placeholder: '',
    query: {},
    requiredLabel: '',
    rows: 5,
    title: '',
    type: 'text',
    value: ''
  };

  state = {
    showError: false,
    value: this.props.value || ''
  };

  componentWillMount () {
    this.updateValue(this.props.query[this.props.name] || this.props.value || '');
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.error) {
      this.showError();
    } else {
      if (!this.needsUpdate(nextProps)) return;
      this.updateValue(
        nextProps.query[this.props.name] ||
        nextProps.value ||
        ''
      );
    }
  }

  // Property initializer binds method to class instance
  onChange = (...args) => {
    this.updateValue(args[0].target.value);
    if (this.props.onChange) this.props.onChange(...args);
    this.hideError();
  };

  needsUpdate ({value, query}) {
    return (
      value !== this.props.value ||
      query !== this.props.query
    );
  }

  updateValue (value) {
    this.setState({value});
  }

  showError () {
    this.setState({showError: true});
  }

  hideError () {
    this.setState({showError: false});
  }

  render () {
    const {
      autoComplete,
      autoFocus,
      disable,
      error,
      id,
      inputStyle,
      label,
      labelHidden,
      maxLength,
      name,
      multiLine,
      pattern,
      placeholder,
      isRequired,
      rows,
      requiredLabel,
      title,
      type
    } = this.props;

    return (
      <div className={`${formGroup} ${styles[inputStyle]}`}>
        {requiredLabel && <span className={`${controlNote} ${controlLabelRequired}`}>{requiredLabel}</span>}
        {labelHidden && <span className={srOnly}>{label}</span>}

        <label
          className={`${controlLabel} ${labelHidden && hidden}`}
          htmlFor={id}>{label}</label>

        <div className={styles.inputContainer}>
          {
            multiLine ?
              <textarea
                autoFocus={autoFocus}
                className={`${formControl} ${styles.textarea}`}
                disabled={disable}
                id={id}
                name={name}
                maxLength={maxLength}
                onChange={this.onChange}
                pattern={pattern}
                placeholder={placeholder}
                required={isRequired}
                rows={rows}
                value={this.state.value} /> :
              <input
                autoComplete={autoComplete}
                autoFocus={autoFocus}
                className={formControl}
                disabled={disable}
                id={id}
                name={name}
                maxLength={maxLength}
                onChange={this.onChange}
                pattern={pattern}
                placeholder={placeholder}
                required={isRequired}
                title={title}
                type={type}
                value={this.state.value} />
          }

          {this.state.showError && error &&
            <span className={errorStyles}>{error}</span>
          }
        </div>
      </div>
    );
  }
}
