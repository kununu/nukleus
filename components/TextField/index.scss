@import '~kununu-theme/scss/base/variables';

.inline {
  position: relative;

  @media (min-width: $screen-sm-min) {
    padding-top: 25px;
  }

  @media (min-width: $screen-xs-min) {
    display: flex;
    flex-flow: row;
    align-items: center;
  }

  label {
    margin-right: 10px;
    margin-bottom: 0;
    width: 140px;
  }
}

.inputContainer {
  position: relative;
  width: 100%;

  * {
    box-sizing: border-box;
  }
}

.textarea {
  resize: none;
}
