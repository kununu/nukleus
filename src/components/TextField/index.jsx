/* eslint "react/no-array-index-key": 0 */

// Read more about controlled components
// https://facebook.github.io/react/docs/forms.html#controlled-components
import React from 'react';
import PropTypes from 'prop-types';
import {queryParamsToObject} from 'utils/params';

import styles from './index.scss';

import Error from '../Error';
import InfoLabel from '../InfoLabel';
import sharedStyles, {
  controlLabel,
  formControl,
  formControlError,
  formGroup,
  hidden,
  srOnly
} from '../index.scss';

export default class TextField extends React.Component {
  static propTypes = {
    autoComplete: PropTypes.string,
    autoFocus: PropTypes.bool,
    disable: PropTypes.bool,
    displayLength: PropTypes.bool,
    dynamicTextareaHeight: PropTypes.bool,
    error: PropTypes.string,
    errorSubInfo: PropTypes.string,
    highlightList: PropTypes.object,
    id: PropTypes.string.isRequired,
    inputStyle: PropTypes.string,
    isRequired: PropTypes.bool,
    label: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.object
    ]).isRequired,
    labelHidden: PropTypes.bool,
    maxLength: PropTypes.number,
    minHeight: PropTypes.number,
    multiLine: PropTypes.bool,
    name: PropTypes.string.isRequired,
    onBlur: PropTypes.func,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    onHighlight: PropTypes.func,
    pattern: PropTypes.string,
    placeholder: PropTypes.string,
    query: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.object
    ]),
    reference: PropTypes.func,
    requiredLabel: PropTypes.string,
    rows: PropTypes.number,
    sanitizeValue: PropTypes.func,
    title: PropTypes.string,
    type: PropTypes.oneOf([
      'email',
      'password',
      'text',
      'url',
      'number',
      'hidden'
    ]),
    value: PropTypes.string
  };

  static defaultProps = {
    autoComplete: 'off',
    autoFocus: false,
    disable: false,
    displayLength: false,
    dynamicTextareaHeight: true,
    error: null,
    errorSubInfo: null,
    highlightList: null,
    inputStyle: 'inline',
    isRequired: false,
    labelHidden: false,
    maxLength: 500,
    minHeight: null,
    multiLine: false,
    onBlur: () => {},
    onChange: null,
    onFocus: () => {},
    onHighlight: () => {},
    pattern: '',
    placeholder: '',
    query: {},
    reference: () => {},
    requiredLabel: '',
    rows: 5,
    sanitizeValue: value => value,
    title: '',
    type: 'text',
    value: ''
  };

  state = {
    highlightedContent: '',
    showError: false,
    textAreaHeight: null,
    value: this.props.value || ''
  };

  componentWillMount () {
    const {query} = this.props;
    const queryObject = queryParamsToObject(query);

    this.updateValue(queryObject[this.props.name] || this.props.value || '');

    // Show error, if already set
    if (this.props.error !== null) this.showError();
  }

  componentWillReceiveProps (nextProps) {
    const queryObject = queryParamsToObject(nextProps.query);

    if (nextProps.error) {
      this.showError();
    } else {
      if (!this.needsUpdate(nextProps)) return;
      this.updateValue(queryObject[this.props.name] ||
        nextProps.value ||
        '');
    }
  }

  // Property initializer binds method to class instance
  onChange = (...args) => {
    const {target} = args[0];
    const {
      dynamicTextareaHeight,
      highlightList,
      multiLine,
      onChange
    } = this.props;

    this.updateValue(this.props.sanitizeValue(target.value));
    if (onChange) this.props.onChange(...args);

    if (dynamicTextareaHeight && multiLine) {
      this.setDynamicTextAreaHeight(target);
    }

    if (highlightList) {
      this.setState({
        highlightedContent: this.getHighlightedContent(target.value)
      });
    }

    this.hideError();
  };

  /**
   * determines which classNames should be added to the label of
   * the component
   *
   * @return {string} [list of classNames split by space]
   */
  get labelClassNames () {
    const {
      inputStyle,
      labelHidden
    } = this.props;
    const classNames = [controlLabel];

    const inputStyles = inputStyle.split(' ');

    // Check if label should be hidden
    if (labelHidden) classNames.push(hidden);

    // Check if TextField contains an error
    if (this.hasError()) classNames.push(sharedStyles.controlLabelError);

    if (inputStyles.indexOf('mediumSize') !== -1) classNames.push(sharedStyles.controlLabelMediumSize);

    return classNames.join(' ');
  }

  /**
   * determines which classNames should be added to the input of
   * the component
   *
   * @return {string} [list of classNames split by space]
   */
  get textFieldClassNames () {
    const {
      multiLine,
      highlightList
    } = this.props;
    const classNames = [formControl];

    // Check if textarea styles need to be added
    if (multiLine) classNames.push(styles.textarea);

    if (multiLine && highlightList) classNames.push(styles.dynamicHeight);

    // Check if error styles need to be added
    if (this.hasError()) classNames.push(formControlError);

    return classNames.join(' ');
  }

  /**
   * determines which classNames should be added to the container of
   * the component
   *
   * @return {string} [list of classNames split by space]
   */
  get containerClassNames () {
    const {
      inputStyle,
      requiredLabel,
      displayLength
    } = this.props;
    const classNames = [formGroup];

    // Add all styles that are added via inputStyles
    const inputStyles = inputStyle.split(' ');
    inputStyles.map(style => classNames.push(sharedStyles[style]));

    if (requiredLabel || displayLength) classNames.push(sharedStyles.paddingTop);

    return classNames.join(' ');
  }

  /**
   * generates the TextField label based on the Textfield label prop
   *
   * @return {ReactElement} [Either returns a label or a react element with the added css class labelContainer]
   */
  get label () {
    const {
      id,
      label
    } = this.props;

    if (typeof label === 'string') {
      return <label className={this.labelClassNames} htmlFor={id}>{label}</label>;
    }

    // We don't simply put a more complex element inside a label to prevent a
    // clickable element like a link or button inside a label
    // However to also add the labelContainer class, we need to return a cloned
    // element and not just the label - element itself
    const classNames = [sharedStyles.labelContainer];

    if (label.props.className) classNames.push(label.props.className);

    return React.cloneElement(
      label,
      {
        ...label.props,
        className: classNames.join(' ')
      }
    );
  }

  getHighlightedContent = contents => {
    const contentRegex = /([^a-zA-Z]+)/;
    const userInputArray = contents.split(contentRegex);

    const {
      highlightList,
      onHighlight
    } = this.props;

    return userInputArray.map((part, i) => {
      if (highlightList[part.toLowerCase()]) {
        onHighlight();
        return <span className={styles.highlighted} key={i}>{part}</span>;
      }
      return part;
    });
  }

  setDynamicTextAreaHeight = target => {
    const oldHeight = Number(this.state.textAreaHeight);
    const currentHeight = target.scrollHeight;

    // Fix so when you backspace it will reduce the correct height
    const newHeight = oldHeight > currentHeight ? currentHeight - 20 : currentHeight;

    if (newHeight > (this.props.minHeight || 134)) {
      this.setState({
        textAreaHeight: newHeight
      });
    }
  }

  needsUpdate ({value, query}) {
    return (
      value !== this.props.value ||
      query !== this.props.query
    );
  }

  updateValue (value) {
    this.setState({value});
  }

  showError () {
    this.setState({showError: true});
  }

  hideError () {
    this.setState({showError: false});
  }

  /**
   * determines which styles should be applied to the textarea
   *
   * @return {object} [object with camelCased properties]
   */
  textAreaStyles = () => {
    const {
      highlightList,
      minHeight
    } = this.props;
    const textAreaStyles = {};

    if (highlightList && minHeight) textAreaStyles.minHeight = minHeight;
    if (this.state.textAreaHeight) textAreaStyles.height = `${this.state.textAreaHeight}px`;

    return textAreaStyles;
  }

  hasError () {
    return this.state.showError && this.props.error;
  }

  render () {
    const {
      autoComplete,
      autoFocus,
      highlightList,
      disable,
      displayLength,
      error,
      errorSubInfo,
      id,
      isRequired,
      label,
      labelHidden,
      maxLength,
      name,
      multiLine,
      onBlur,
      onFocus,
      pattern,
      placeholder,
      reference,
      requiredLabel,
      rows,
      title,
      type
    } = this.props;

    const {highlightedContent} = this.state;

    return (
      <div className={this.containerClassNames} id={`${name}-container`}>
        {this.state.highlightContent}
        <InfoLabel
          requiredLabel={requiredLabel}
          inputValue={this.state.value}
          displayLength={displayLength}
          maxLength={maxLength} />
        {labelHidden && <span className={srOnly}>{label}</span>}

        {this.label}

        <div className={styles.inputContainer}>
          {highlightList ?
            <div className={`${styles.highlightOverlay} ${this.textFieldClassNames}`}>
              {highlightedContent}
            </div>
            : null
          }
          {
            multiLine ?
              <textarea
                autoFocus={autoFocus}
                className={this.textFieldClassNames}
                disabled={disable}
                id={id}
                name={name}
                maxLength={maxLength}
                onChange={this.onChange}
                onBlur={onBlur}
                onFocus={onFocus}
                pattern={pattern}
                placeholder={placeholder}
                required={isRequired}
                ref={reference}
                rows={rows}
                style={this.textAreaStyles()}
                value={this.state.value} /> :
              <input
                autoComplete={autoComplete}
                autoFocus={autoFocus}
                className={this.textFieldClassNames}
                disabled={disable}
                id={id}
                name={name}
                maxLength={maxLength}
                onChange={this.onChange}
                onBlur={onBlur}
                onFocus={onFocus}
                pattern={pattern}
                placeholder={placeholder}
                ref={reference}
                required={isRequired}
                title={title}
                type={type}
                value={this.state.value} />
          }

          {this.hasError() &&
            <Error
              info={error}
              subInfo={errorSubInfo} />
          }
        </div>
      </div>
    );
  }
}
