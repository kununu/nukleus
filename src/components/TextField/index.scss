@import '~@kununu/kununu-theme-v2/scss/base/variables';

.inputContainer {
  position: relative;
  width: 100%;

  * {
    box-sizing: border-box;
  }
}

.textarea {
  resize: none;

  &.dynamicHeight {
    min-height: 134px;
    overflow-y: hidden;
  }
}

.highlightOverlay {
  margin-bottom: 0;
  position: absolute;
  z-index: 40;
  top: 1px;
  left: 1px;
  bottom: 0;
  height: auto;
  pointer-events: none;
  overflow: hidden;
  padding-bottom: 30px;
  word-wrap: break-word;
  white-space: pre-wrap;
  text-align: start;
  vertical-align: text-bottom;
  border: 0;
  background-color: transparent;
  color: transparent;

  & + input,
  & + textarea {
    margin: 0;
    position: relative;
    background-color: transparent;
  }
}

.highlighted {
  background-color: rgba($red, .1);
  padding: 2px;
  top: 2px;
  margin: -3px;
  line-height: 0;
  position: relative;
  border-radius: 2px;
  border: 1px solid rgba($red, .15);
}

.highlightOverlay.textarea {
  .highlighted {
    position: static;
  }
}
