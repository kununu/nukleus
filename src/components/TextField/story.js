import React from 'react';
import {storiesOf} from '@storybook/react';
import {defaultOptions} from 'defaultOptions';
import {withOptions} from '@storybook/addon-options';
// import {action} from '@storybook/addon-actions';
// import {withNotes} from '@storybook/addon-notes';
// import {withInfo} from '@storybook/addon-info';
// import {withKnobs, text, boolean} from '@storybook/addon-knobs/react';
// import backgrounds from '@storybook/addon-backgrounds';

import TextField from './index';

storiesOf('TextField', module)
  .addDecorator(withOptions({...defaultOptions}))
  .add('with some emoji', () => (
    <TextField
      id="text-field"
      label="TextField"
      inputStyle="mediumSize"
      highlightList={{bad: 'bad'}}
      multiLine
      minHeight={80}
      name="text-field" />
  ));
