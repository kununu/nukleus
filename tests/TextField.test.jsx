import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import toJson from 'enzyme-to-json';
import TextField from 'TextField'; // eslint-disable-line import/no-unresolved, import/extensions, import/no-extraneous-dependencies

const textField = (
  <TextField
    id="text-field"
    label="TextField"
    name="text-field"
    value="test"
    isRequired
    requiredLabel="required" />
);

// input
test('Renders TextField without crashing', () => {
  const component = renderer.create(textField);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Changes output on value change manipulation', () => {
  const component = mount(textField);
  component.find('input').simulate('change', {target: {value: 'UGH'}});
  expect(toJson(component)).toMatchSnapshot();
});

// textarea
test('Renders TextField with multiline without crashing', () => {
  const component = renderer.create(
    <TextField
      id="text-field"
      label="TextField"
      name="text-field"
      value="test2"
      multiLine
      isRequired
      requiredLabel="required" />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
