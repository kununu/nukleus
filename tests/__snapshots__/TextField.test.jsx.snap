exports[`test Changes output on value change manipulation 1`] = `
<TextField
  autoComplete="off"
  autoFocus={false}
  disable={false}
  error={null}
  id="text-field"
  inputStyle="inline"
  isRequired={true}
  label="TextField"
  labelHidden={false}
  maxLength={500}
  multiLine={false}
  name="text-field"
  onChange={null}
  pattern=""
  placeholder=""
  query={Object {}}
  requiredLabel="required"
  rows={5}
  title=""
  type="text"
  value="test">
  <div
    className="formGroup inline">
    <span
      className="controlNote controlLabelRequired">
      required
    </span>
    <label
      className="controlLabel false"
      htmlFor="text-field">
      TextField
    </label>
    <div
      className="inputContainer">
      <input
        autoComplete="off"
        autoFocus={false}
        className="formControl"
        disabled={false}
        id="text-field"
        maxLength={500}
        name="text-field"
        onChange={[Function]}
        pattern=""
        placeholder=""
        required={true}
        title=""
        type="text"
        value="UGH" />
    </div>
  </div>
</TextField>
`;

exports[`test Renders TextField with multiline without crashing 1`] = `
<div
  className="formGroup inline">
  <span
    className="controlNote controlLabelRequired">
    required
  </span>
  <label
    className="controlLabel false"
    htmlFor="text-field">
    TextField
  </label>
  <div
    className="inputContainer">
    <textarea
      autoFocus={false}
      className="formControl textarea"
      disabled={false}
      id="text-field"
      maxLength={500}
      name="text-field"
      onChange={[Function]}
      pattern=""
      placeholder=""
      required={true}
      rows={5}
      value="test2" />
  </div>
</div>
`;

exports[`test Renders TextField without crashing 1`] = `
<div
  className="formGroup inline">
  <span
    className="controlNote controlLabelRequired">
    required
  </span>
  <label
    className="controlLabel false"
    htmlFor="text-field">
    TextField
  </label>
  <div
    className="inputContainer">
    <input
      autoComplete="off"
      autoFocus={false}
      className="formControl"
      disabled={false}
      id="text-field"
      maxLength={500}
      name="text-field"
      onChange={[Function]}
      pattern=""
      placeholder=""
      required={true}
      title=""
      type="text"
      value="test" />
  </div>
</div>
`;
